### PR TITLE
octopus: bluestore: fix collection_list ordering

### PR DESCRIPTION
**Pull request by trociny**
backport tracker: https://tracker.ceph.com/issues/47363

---

backport of https://github.com/ceph/ceph/pull/36230
parent tracker: https://tracker.ceph.com/issues/43174

this backport was staged using ceph-backport.sh version 15.1.1.389
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -686,26 +686,99 @@ namespace {
...
+
+  void cache_key() {
+    if (!valid()) {
+      return;
+    }
 
-  bool valid() const {
+    m_key = m_it->key();
+    int r = get_key_object(m_key, &m_oid);
+    ceph_assert(r != -1);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Shouldn't this be >= 0?
get_key_object might return -2 through -7 as well

**Review by ifed01 — APPROVED** [review]

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc]: It was changed in the "refactoring" commit anyway.

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: right, sorry, forget to remove this comment

**Comment by yuriw** [comment]
[wip-yuri7-testing-2020-10-04-0852-octopus](https://trello.com/c/tLpholMk/1031-wip-yuri7-testing-2020-10-04-0852-octopus)

**Pull request merged**